src/shared/userdb.c

```diff
@@ -159,6 +171,72 @@ static void membership_data_done(struct membership_data *d) {
         free(d->group_name);
 }
 
+static int userdb_maybe_restart_query(
+                UserDBIterator *iterator,
+                sd_varlink *link,
+                sd_json_variant *parameters,
+                const char *error_id) {
+
+        int r;
+
+        assert(iterator);
+        assert(link);
+        assert(error_id);
+
+        /* These fields were added in v258 and didn't exist in previous implementations. Hence, we consider
+         * their support optional: if any service refuses any of these fields, we'll restart the query
+         * without them, and apply the filtering they are supposed to do client side. */
+        static const char *const fields[] = {
+                "fuzzyNames",
+                "dispositionMask",
+                "uidMin",
+                "uidMax",
+                "gidMin",
+                "gidMax",
+                NULL
+        };
+
+        /* Figure out if the reported error indicates any of the suppressable fields are at fault, and that
+         * our query actually included them */
+        bool restart = false;
+        STRV_FOREACH(f, fields) {
+                if (!sd_varlink_error_is_invalid_parameter(error_id, parameters, *f))
+                        continue;
+
+                if (!sd_json_variant_by_key(iterator->query, *f))
+                        continue;
+
+                restart = true;
+                break;
+        }
+
+        if (!restart)
+                return 0;
+
+        /* Now patch the fields out */
+        _cleanup_(sd_json_variant_unrefp) sd_json_variant *patched_query =
+                sd_json_variant_ref(iterator->query);
+
+        STRV_FOREACH(f, fields) {
+                r = sd_json_variant_unset_field(&patched_query, *f);
```

Comment:
BTW, maybe better to introduce `_unset_fields()` which takes strv ?

Reply:
oh! i am such an idiot! now that you say that I realize this already exists, and is called sd_json_variant_filter(), and sd_json_variant_unset_field() is just a dumb mistake and should be removed again, because entirely redundant